jax/experimental/fused_attention_stableHLO.py

```diff
@@ -0,0 +1,561 @@
+    out_bdims = query_bdim, query_bdim, query_bdim
+
+    *batch_tuple, q_seq_len, num_heads, head_dim = query.shape
+    *_, kv_seq_len, _, _ = key.shape
+    batch = reduce(operator.mul, batch_tuple)
+    # reshape to 4D shape
+    query_reshape = jnp.reshape(query, (batch, q_seq_len, num_heads, head_dim))
+    key_reshape = jnp.reshape(key, (batch, kv_seq_len, num_heads, head_dim))
+    value_reshape = jnp.reshape(value, (batch, kv_seq_len, num_heads, head_dim))
+    bias_reshape = jnp.reshape(bias, (batch, num_heads, q_seq_len, kv_seq_len))
+    mask_reshape = jnp.reshape(mask, (batch, num_heads, q_seq_len, kv_seq_len))
+    activation_reshape = jnp.reshape(activation, (batch, num_heads, q_seq_len, kv_seq_len))
+    fwd_output_reshape = jnp.reshape(fwd_output, (batch, q_seq_len, num_heads, head_dim))
+    grad_output_reshape = jnp.reshape(grad_output, (batch, q_seq_len, num_heads, head_dim))
+
+    grad_query, grad_key, grad_value = _dot_product_attention_bwd_p_outer.bind(
+        query_reshape, key_reshape, value_reshape, bias_reshape,
+        mask_reshape, activation_reshape, fwd_output_reshape, grad_output_reshape,
+        scale=scale, seed=seed, dropout_rate=dropout_rate,
+        variadic_args=variadic_args, is_flash_attention=is_flash_attention,
+        is_causal_mask=is_causal_mask)
+    
+    # reshape to original shape
+    grad_query = jnp.reshape(grad_query, (*batch_tuple, q_seq_len, num_heads, head_dim))
+    grad_key = jnp.reshape(grad_key, (*batch_tuple, kv_seq_len, num_heads, head_dim))
+    grad_value = jnp.reshape(grad_value, (*batch_tuple, kv_seq_len, num_heads, head_dim))
+    grads = (grad_query, grad_key, grad_value)
+    return grads, out_bdims
+
+# custom partitioning
+def get_padded_spec(arg_info):
+    spec = None if arg_info.sharding is None else arg_info.sharding.spec
+    ndim = arg_info.ndim
+    if spec is None:
+        return (None,) * ndim
+    assert len(spec) <= ndim
+    return spec + (None,) * (ndim - len(spec))
+    
+# fwd custom partition
+_dot_product_attention_fwd_lower = custom_partitioning(_dot_product_attention_fwd_impl, static_argnums=(5,6,7,8,9,10))
+def _dot_product_attention_fwd_infer_sharding_from_operands(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, mesh, arg_shapes, result_shape):
+    # (*batch, q_seq, num_head, head)
+    query_spec = get_padded_spec(arg_shapes[0])
+    # (*batch, kv_seq, num_head, head)
+    key_spec = get_padded_spec(arg_shapes[1])
+    # keep out sharding same as query sharding since they have same shape
+    out_sharding = NamedSharding(mesh, PartitionSpec(*query_spec))
+    # activation sharding
+    activation_sharding = NamedSharding(mesh, PartitionSpec(*query_spec[:-3], query_spec[-2], query_spec[-3], key_spec[-3]))
+    return (out_sharding, activation_sharding)
+
+def _dot_product_attention_fwd_partition(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, mesh, arg_shapes, result_shape):
```

Comment:
In this function and the one above, you don't force any sharding. Like if all inputs sharding are supported.
I'm pretty sure the kernel don't support all sharding.
You can inspire yourself from the equivalent in TE:
https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/cpp_extensions.py#L1822-L1851

Reply:
Not exactly sure what you meant here. Could you add an example?

Reply:
Let say the kernel know how to shard only on the batch dimensions.
Now, maybe the inputs will be sharded on the inner dimensions or the sequence dimensions.
So you need to force the replication on all dimensions except the batch dimensions.
This is done by using None in TE.